src/future.ml

```diff
@@ -350,30 +367,100 @@ module Scheduler = struct
       ~output:output
       ~exit_status:status;
     if not exiting then begin
+      let pp_purpose ppf = function
+      | Internal_job ->
+         Format.fprintf ppf "(internal)"
+      | Build_job (targets, build_dir, context_name) ->
+         let strip_prefix path =
```

Comment:
Done. I didn't know about Path.extract_build_context, it simplifies things a lot. Now Futures need only be passed the target list, not extra context info.